### PR TITLE
[Inductor][CPP] Enable Quantized Linear with AMX MicroGEMM

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129470
* #129221
* __->__ #129220
* #129103
* #129049
* #129048
* #128825

**Summary**
Add the AMX micro gemm kernel with int8 data type.

**Test Plan**
```
clear && python -u -m pytest -s -v test/inductor/test_cpu_select_algorithm.py -k test_quantized_linear_amx
```

**Next Step**
- [✓] Unary post op fusion
- [✓] Int8 output
- [✓] Binary Fusion
- [✓] AMX int8 MicroGEMM Kernel

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/inductor/test_cpu_select_algorithm.py [hunk 1]
```diff
@@ -523,6 +523,38 @@ def forward(self, x, other, other2):
...
+        )
+        atol, rtol = 1e-2, 1e-2
+        with patch.object(select_algorithm, "VERIFY", dict(atol=atol, rtol=rtol)):
+            self.common(ref_quantized_mod, (v,), atol=atol, rtol=rtol)
+        self.assertEqual(counters["inductor"]["select_algorithm_autotune"], 1)
+        vec_amx = VecAMX()
+        if vec_amx:
+            self.assertTrue(counters["inductor"]["cpp_micro_gemm_amx_counter"] > 0)
+        else:
+            self.assertEqual(counters["inductor"]["cpp_micro_gemm_amx_counter"], 0)
```
Comment (jgong5) [inline test/inductor/test_cpu_select_algorithm.py:556]: refactor this into a util checking function?

File: torch/_inductor/codegen/cpp_micro_gemm.py [hunk 2]
```diff
@@ -553,9 +566,17 @@ class CppMicroGemmAMX(CppMicroGemm):
         _tile_loadd({{tile_idx_a}}, A + {{tile_row * 16}} * lda + k, lda * sizeof({{input_t}}));
         {%- endif %}
         {%- if tile_row == 0 %}
+        {%- if int8_gemm %}
+        _tile_loadd({{tile_idx_b}}, B + k * ldb + {{tile_col * 16 * 4}}, ldb * 4 * sizeof({{input_t}}));
+        {%- else %}
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_micro_gemm.py:571]: Can be simplified with "vnni_size"?

File: torch/_inductor/codegen/cpp_micro_gemm.py [hunk 3]
```diff
@@ -677,7 +704,11 @@ def create_from_config(cls, config: CppMicroGemmConfig):
                 continue
             if (
                 config.input_dtype == input_dtype
-                and config.output_dtype == output_dtype
+                and (
+                    config.output_dtype == torch.int32
+                    if input_dtype == torch.uint8
+                    else config.output_dtype == output_dtype
+                )
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_micro_gemm.py:711]: Is it better to check this from the caller side? The caller makes sure that the `output_dtype` is torch.int32 in the case of int8 gemm?

File: torch/_inductor/codegen/cpp_micro_gemm.py [hunk 4]
```diff
@@ -593,7 +614,10 @@ def codegen_define(self, kernel: CppTemplateKernel) -> str:
         block_m, block_n, block_k = self.register_blocking
         assert block_m % 16 == 0, "Only support block_m % 16 == 0 for AMX"
         assert block_n % 16 == 0, "Only support block_n % 16 == 0 for AMX"
-        assert block_k == 32, "Only support block_k = 32 for AMX"
+        if self.input_dtype == torch.uint8:
+            assert block_k == 64, "Only support block_k = 64 for AMX INT8"
+        else:
+            assert block_k == 32, "Only support block_k = 32 for AMX Half"
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_micro_gemm.py:620]:
```suggestion
            assert block_k == 32, "Only support block_k = 32 for AMX Bfloat16/Float16"
```

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp_micro_gemm.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_micro_gemm.py:571]: Sure, added a render option of `vnni_size`

File: test/inductor/test_cpu_select_algorithm.py (hunk 1 above)
Comment (leslie-fang-intel) [reply test/inductor/test_cpu_select_algorithm.py:556]: Sure, added a util function to check AMX usage.

File: torch/_inductor/codegen/cpp_micro_gemm.py (hunk 4 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_micro_gemm.py:620]: Thanks and changed.

File: torch/_inductor/codegen/cpp_micro_gemm.py (hunk 3 above)
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_micro_gemm.py:711]: Thanks for the comment. I think this line of change is not needed anymore. Remove this change in this PR.

File: test/inductor/test_cpu_select_algorithm.py [hunk 5]
```diff
@@ -358,10 +366,7 @@ def forward(self, x):
             self.common(mod, (v,), atol=atol, rtol=rtol)
         self.assertEqual(counters["inductor"]["select_algorithm_autotune"], 1)
         vec_amx = VecAMX()
-        if vec_amx:
-            self.assertTrue(counters["inductor"]["cpp_micro_gemm_amx_counter"] > 0)
-        else:
-            self.assertEqual(counters["inductor"]["cpp_micro_gemm_amx_counter"], 0)
+        self._check_amx_usage(vec_amx)
```
Comment (jgong5) [inline test/inductor/test_cpu_select_algorithm.py:369]:
```suggestion
        self._check_amx_counter(vec_amx)
```

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**